File: app/src/main/java/com/example/home/paging/MainPagingSource.java
package com.example.home.paging;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.paging.PagingSource;
import androidx.paging.PagingState;
import androidx.paging.rxjava3.RxPagingSource;

import com.example.home.model.HitsItem;
import com.example.home.model.Page;
import com.example.home.network.RetroInstance;
import com.example.home.utils.Constants;

import java.util.List;

import io.reactivex.rxjava3.core.Single;

public class MainPagingSource extends RxPagingSource<Integer , HitsItem> {
    @Nullable
    @Override
    public Integer getRefreshKey(@NonNull PagingState<Integer, HitsItem> pagingState) {
        Integer anchorPosition = pagingState.getAnchorPosition();
        if (anchorPosition == null) {
            return null;
        }

        LoadResult.Page<Integer, HitsItem> anchorPage = pagingState.closestPageToPosition(anchorPosition);
        if (anchorPage == null) {
            return null;
        }

        Integer prevKey = anchorPage.getPrevKey();
        if (prevKey != null) {
            return prevKey + 1;
        }

        Integer nextKey = anchorPage.getNextKey();
        if (nextKey != null) {
            return nextKey - 1;
        }
        return pagingState.getAnchorPosition();
    }

    @NonNull
    @Override
    public Single<LoadResult<Integer, HitsItem>> loadSingle(@NonNull LoadParams<Integer> loadParams) {

        try{
            int pageNumber = loadParams.getKey() != null? loadParams.getKey():1;

            return RetroInstance.getRetroService()
                    .getRemotePage(Constants.URL_TAG,pageNumber)
                    .map(Page::getHits)
                    .map(hitsItems -> toLoadResult(hitsItems,pageNumber))
                    .onErrorReturn(LoadResult.Error::new);

        }catch (Exception e){
            return Single.just(new LoadResult.Error<>(e));
        }
    }

    private LoadResult<Integer , HitsItem> toLoadResult(List<HitsItem> hitsItems, int pageNumber) {
        return new LoadResult.Page(hitsItems,
                pageNumber == 1?null:pageNumber-1,
                pageNumber == 50?null:pageNumber+1,
                LoadResult.Page.COUNT_UNDEFINED,
                LoadResult.Page.COUNT_UNDEFINED);
    }
}

File: app/src/main/java/com/example/assignment13_mvvm_kotlin/pagingSource/MainPagingSource.kt
package com.example.assignment13_mvvm_kotlin.pagingSource

import androidx.paging.PagingSource
import androidx.paging.PagingState
import com.example.assignment13_mvvm_kotlin.model.HitsItem
import com.example.assignment13_mvvm_kotlin.model.Page
import com.example.assignment13_mvvm_kotlin.network.RetroService
import com.example.assignment13_mvvm_kotlin.utils.Constants
import javax.inject.Inject

class MainPagingSource @Inject constructor(val service: RetroService ) : PagingSource<Int , HitsItem>() {


    override fun getRefreshKey(state: PagingState<Int, HitsItem>): Int? {
        return state.anchorPosition?.let { anchorPosition ->
            state.closestPageToPosition(anchorPosition)?.prevKey?.plus(1)
                ?: state.closestPageToPosition(anchorPosition)?.nextKey?.minus(1)
        }
    }

    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, HitsItem> {
        return try {

            val currentPage : Int = params.key?:Constants.FIRST_PAGE_NUMBER
            var nextPage : Int? = null
            var prevPage : Int? = null

            val page = service.getRemotePage(Constants.URL_TAG,currentPage)

            nextPage = if (page.nbPages!! > currentPage){
                currentPage+1
            }else{
                null
            }
            prevPage = if(currentPage==1)null else currentPage-1

            LoadResult.Page(data = page.hits!!  , nextKey = nextPage , prevKey = prevPage)
        }
        catch (e : Exception){
            LoadResult.Error(e)
        }
    }
}